Rep crawler overhaul

pwojcikdev opened the pull request:
Initial motivation for this PR was providing better control over rep crawler component lifetime and running its internal logic on a separate dedicated thread. In the process of working on it I made many smaller improvements that should make the logic clearer and less error prone.

Testing results provided by @gr0vity-dev show that performance is on par with current develop.
![image](https://github.com/nanocurrency/nano-node/assets/3044353/60279544-5cde-4db6-b962-e56723b2fafd)

pwojcikdev committed:
Organize members

pwojcikdev committed:
Single mutex

pwojcikdev committed:
Move collect_container_info into source file

pwojcikdev committed:
Overhaul `rep_crawler`

pwojcikdev committed:
Move rep_crawler tests to a dedicated file

pwojcikdev committed:
Improve tests

pwojcikdev committed:
Simplify `nano::representative` struct

pwojcikdev committed:
Rework `rep_crawler::process ()`

pwojcikdev committed:
Use pointer as channel identity

pwojcikdev committed:
Move remaining rep_crawler tests

pwojcikdev committed:
Overhaul `rep_crawler` part 2

pwojcikdev committed:
Improve channel logging

pwojcikdev committed:
Fix tests

pwojcikdev committed:
Cleanup

pwojcikdev committed:
Uniform rep crawler log enum naming

pwojcikdev committed:
Notify condition after successful response

pwojcikdev committed:
Allow multiple queries per channel during warmup

pwojcikdev committed:
Simplify `keepalive_preconfigured ()`

pwojcikdev committed:
Improve `collect_container_info ()`

pwojcikdev committed:
No limiter drop

pwojcikdev committed:
Adjustments

pwojcikdev commented:
I made a few adjustments based on recent network stress test. It should make repcrawler a bit better at finding representatives when vote requests are unreliable.
##### develop
<img width="2040" alt="Screenshot 2024-02-28 at 13 48 37" src="https://github.com/nanocurrency/nano-node/assets/3044353/5ac70004-2433-4ded-bad4-a3f729e37103">

##### rep-crawler-overhaul
<img width="2041" alt="Screenshot 2024-02-28 at 13 48 30" src="https://github.com/nanocurrency/nano-node/assets/3044353/40f90b4b-4c18-4936-b5d1-8d601a2ac6da">

pwojcikdev committed:
Throttle queries to active reps

dsiganos commented on nano/lib/config.hpp:
```diff
@@ -307,6 +309,9 @@ class network_constants
 	std::chrono::seconds optimistic_activation_delay{ 30 };
 
+	std::chrono::milliseconds rep_crawler_normal_interval{ 1000 * 7 };
+	std::chrono::milliseconds rep_crawler_warmup_interval{ 1000 * 3 };
```
We know that too many confirm_req are not a good thing.
If the network is in trouble and it has no sufficient weight, do we really want to flood with confirm requests?
I think we should remove this setting all-together.

pwojcikdev replied:
I disagree. I believe the benefits of having a quick warmup phase (finding principal representatives quickly) outweigh minimally higher resource consumption. At the peak, rep crawler published ~20-25 confirm reqs per second and that phase took a fraction of a second. It then quickly throttled down.

25 requests per second pales in comparison to confirm reqs being published by active elections, which average 3k-4k per second, and they are the main bottleneck.

<img width="1436" alt="Screenshot 2024-03-05 at 14 35 10" src="https://github.com/nanocurrency/nano-node/assets/3044353/3e7db7a9-288e-4150-9b4e-45308d0da919">

pwojcikdev commented:
With the latest commit running on live network, rep crawler averages about ~6 requests per second with response rate of about 25%. It is able to keep a stable peered stake.
<img width="1277" alt="Screenshot 2024-03-05 at 14 23 09" src="https://github.com/nanocurrency/nano-node/assets/3044353/45186f69-058e-4a57-b826-ea3163d2c5eb">

dsiganos commented on nano/lib/config.hpp:
```diff
@@ -307,6 +309,9 @@ class network_constants
 	/** How much to delay activation of optimistic elections to avoid interfering with election scheduler */
 	std::chrono::seconds optimistic_activation_delay{ 30 };
 
+	std::chrono::milliseconds rep_crawler_normal_interval{ 1000 * 7 };
```
I think 7 seconds is excessive. How about 30 seconds?

dsiganos commented on nano/lib/config.hpp:
```diff
@@ -251,6 +251,8 @@ class network_constants
 			telemetry_broadcast_interval = 500ms;
 			optimistic_activation_delay = 2s;
+			rep_crawler_normal_interval = 500ms;
+			rep_crawler_warmup_interval = 500ms;
```
These used to be 100ms, what is the reason for changing them?
Seems like a large difference.

pwojcikdev replied:
Since the rep crawler is now logging to debug, decreasing it will lead to more noise in logs. If there is a noticeable impact on unit test runtimes we may want to tune it, but I would keep this parameter conservative.

dsiganos commented on nano/node/repcrawler.hpp:
```diff
@@ -52,103 +49,143 @@ class representative
 	 * @return false if any vote passed the checks and was added to the response queue of the rep crawler
 	 */
-	bool response (std::shared_ptr<nano::transport::channel> const &, std::shared_ptr<nano::vote> const &, bool force = false);
+	bool process (std::shared_ptr<nano::vote> const &, std::shared_ptr<nano::transport::channel> const &);
```
The comment above this function is now wrong. The return value has reverse polarity and there is no force arg,

dsiganos commented on nano/node/repcrawler.cpp:
```diff
@@ -1,370 +1,524 @@
 
 				// Update if representative channel was changed
-				if (info.channel->get_endpoint () != channel->get_endpoint ())
+				if (rep.channel->get_endpoint () != channel->get_endpoint ())
```
Technically, it would be better to compare channel rather than endpoints.
It is possible two have two channels with the same remote endpoint, if the other computer suddenly reboots.
This will also work because eventually the old channel should timeout but it will take longer to notice.

pwojcikdev replied:
We initiate two channels between peers, one in, one out. Isn't comparing channels going to lead to excessive channel switching? I assume the original author of the code used this comparison to fix this problem.

dsiganos replied:
Hmmm, maybe that is the reason. OK.

dsiganos commented on nano/node/repcrawler.cpp:
```diff
@@ -1,370 +1,524 @@
 		{
-			probable_reps.modify (existing, [rep_weight, &updated, &vote, &channel, &prev_channel] (nano::representative & info) {
-				info.last_response = std::chrono::steady_clock::now ();
+			reps.modify (existing, [rep_weight, &updated, &vote, &channel, &prev_channel] (rep_entry & rep) {
```
rep_weight is not used

dsiganos commented on nano/node/repcrawler.cpp:
```diff
@@ -1,370 +1,524 @@
+
+	// TODO: Make these values configurable
+	constexpr std::size_t conservative_count = 160;
+	constexpr std::size_t aggressive_count = 160;
```
The conservative count is the same as the aggressive count. Is this a mistake?
Also, these numbers are much bigger than the previous code.

dsiganos commented:
I am trying to understand how multiple reps inside one node would be dealt.
So I created this test case and it fails:
```
// Test that nodes can track PRs when multiple PRs are inside one node
TEST (rep_crawler, two_reps_one_node)
{
	nano::test::system system;
	auto & node1 = *system.add_node ();
	auto & node2 = *system.add_node ();

	// create a second PR account
	nano::keypair pr1 = nano::test::setup_rep (system, node1, node1.balance (nano::dev::genesis_key.pub) / 10);

	ASSERT_EQ (0, node2.rep_crawler.representative_count ());

	// enable the two PRs in node1
	system.wallet (0)->insert_adhoc (nano::dev::genesis_key.prv);
	system.wallet (0)->insert_adhoc (pr1.prv);

	ASSERT_TIMELY_EQ (5s, node2.rep_crawler.representative_count (), 2);
	auto reps = node2.rep_crawler.representatives ();
	ASSERT_EQ (2, reps.size ());
}
```

pwojcikdev commented:
@dsiganos Good catch, feel free to commit this testcase. I'll look into solving it.

dsiganos commented on nano/node/repcrawler.cpp:
```diff
@@ -1,370 +1,524 @@
+			// TODO: Track query response time
+
+			responses.push_back ({ channel, vote });
+			queries.erase (it);
```
The query cannot be erased here; if a node has multiple PRs served by one channel then votes after the first will not be processed by the rep crawler due to the query being deleted by the first vote.

pwojcikdev replied:
Yeah, I have a pretty good idea on how to fix this. In the meantime it would be great if you could add your testcase to the branch.

dsiganos replied:
We should not be actively deleting the query because we cannot possibly know when to delete it.
We should just let it timeout.

pwojcikdev replied:
Yes, this is an easy and obvious fix.

dsiganos committed:
Rep crawler test case for having 2 reps inside one node

dsiganos committed:
rep_query_interval needs to be faster for dev network

dsiganos committed:
Fix rep tracking when there are multiple reps inside one node

Do not delete the query when a confirm_ack is received with the right vote.
We might get more replies and we have no way of knowing how many will come.
Just let the query timeout.

dsiganos committed:
Do not complain when a query timeout, if replies were received

dsiganos committed:
Merge branch 'develop' into repcrawler-overhaul-2

# Conflicts:
#	nano/lib/stats_enums.hpp
#	nano/lib/thread_roles.cpp
#	nano/lib/thread_roles.hpp

dsiganos committed:
Fix comment of rep_crawler::process function
